_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,231 @@
+as before. However, if it's a test results file (eg: .trx), one should call the
+`POST /build/{buildid}/test-results` endpoint instead. This tells the server
+to parse them as well, creating several `TestResultDetail`s and one `TestResultSummary`
+per TRX file.
+
+The summaries get inserted into the database table `test_result_summary`.\
+The results get inserted into the database table `test_result_detail`.
+
+![Database structure](../_assets/wharf-db-graph.png)
+
+#### New or modified database models
+```diff
+// database_models.go
+type Build struct {
+  BuildID     uint         `gorm:"primaryKey" json:"buildId"`
+  StatusID    BuildStatus  `gorm:"not null" json:"statusId"`
+  ProjectID   uint         `gorm:"not null;index:build_idx_project_id" json:"projectId"`
+  Project     *Project     `gorm:"foreignKey:ProjectID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+  ScheduledOn *time.Time   `gorm:"nullable;default:NULL" json:"scheduledOn" format:"date-time"`
+  StartedOn   *time.Time   `gorm:"nullable;default:NULL" json:"startedOn" format:"date-time"`
+  CompletedOn *time.Time   `gorm:"nullable;default:NULL" json:"finishedOn" format:"date-time"`
+  GitBranch   string       `gorm:"size:300;default:'';not null" json:"gitBranch"`
+  Environment null.String  `gorm:"nullable;size:40" json:"environment" swaggertype:"string"`
+  Stage       string       `gorm:"size:40;default:'';not null" json:"stage"`
+  Params      []BuildParam `gorm:"foreignKey:BuildID" json:"params"`
+  IsInvalid   bool         `gorm:"not null;default:false" json:"isInvalid"` 
++ TestResultSummaries []TestResultSummary `gorm:"foreignKey:BuildID" json:"testResultSummaries"`
+}
+
++type TestResultSummary struct {
++  ArtifactID  uint      `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
++  Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Total       uint      `gorm:"not null" json:"total"`
++  Failed      uint      `gorm:"not null" json:"failed"`
++  Passed      uint      `gorm:"not null" json:"passed"`
++  Skipped     uint      `gorm:"not null" json:"skipped"`
++}
+
++type TestResultDetail struct {
++  ArtifactID  uint              `gorm:"not null;index:testresult_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact         `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Name        string	        `gorm:"not null" json:"name"`
++  Message     string            `gorm:"nullable" json:"message"`
++  StartedOn   *time.Time        `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
++  CompletedOn *time.Time        `gorm:"nullable;default:NULL;" json:"finishedOn" format:"date-time"`
```

Comment:
nit: we ought to be using the `null` package for nullable fields instead of pointers.

Also, inconsistent json property name.

```suggestion
+  StartedOn   null.Time         `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
+  CompletedOn null.Time         `gorm:"nullable;default:NULL;" json:"completedOn" format:"date-time"`
```

Reply:
Maybe deserves its own RFC regarding the `null.Time` as we've used `*time.Time` elsewhere. You're free to resolve with only the json change:

```suggestion
+  StartedOn   *time.Time        `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
+  CompletedOn *time.Time        `gorm:"nullable;default:NULL;" json:"completedOn" format:"date-time"`
```